File: android/app/src/main/java/deckers/thibault/aves/utils/PathSegments.java
package deckers.thibault.aves.utils;

import androidx.annotation.NonNull;

import java.io.File;

public class PathSegments {
    private String storage;
    private String relativePath;
    private String filename;

    public PathSegments(@NonNull String path, @NonNull String[] storageVolumePaths) {
        for (int i = 0; i < storageVolumePaths.length && storage == null; i++) {
            if (path.startsWith(storageVolumePaths[i])) {
                storage = storageVolumePaths[i];
            }
        }

        int lastSeparatorIndex = path.lastIndexOf(File.separator) + 1;
        if (lastSeparatorIndex > storage.length()) {
            filename = path.substring(lastSeparatorIndex);
            relativePath = path.substring(storage.length(), lastSeparatorIndex);
        }
    }

    public String getStorage() {
        return storage;
    }

    public String getRelativePath() {
        return relativePath;
    }

    public String getFilename() {
        return filename;
    }
}

File: android/app/src/main/kotlin/deckers/thibault/aves/storage/PathSegments.kt
package deckers.thibault.aves.storage

import android.content.Context
import deckers.thibault.aves.model.FieldMap
import deckers.thibault.aves.storage.StorageUtils.ensureTrailingSeparator
import deckers.thibault.aves.storage.StorageUtils.getVolumePath
import java.io.File

// `fullPath` should match "volumePath + relativeDir + fileName"
class PathSegments {
    var volumePath: String? = null // `volumePath` with trailing "/"
    var relativeDir: String? = null // `relativeDir` with trailing "/"
    private var fileName: String? = null // null for directories

    constructor(context: Context, fullPath: String) {
        volumePath = getVolumePath(context, fullPath)
        if (volumePath != null) {
            val lastSeparatorIndex = fullPath.lastIndexOf(File.separator) + 1
            val volumePathLength = volumePath!!.length
            if (lastSeparatorIndex > volumePathLength) {
                fileName = fullPath.substring(lastSeparatorIndex)
                relativeDir = fullPath.substring(volumePathLength, lastSeparatorIndex)
            }
        }
    }

    constructor(volumePath: String?, relativeDir: String?) {
        this.volumePath = volumePath
        this.relativeDir = if (relativeDir != null) ensureTrailingSeparator(relativeDir) else null
    }

    override fun toString(): String = "PathSegments#${hashCode()}{volumePath=$volumePath relativeDir=$relativeDir fileName=$fileName}"

    fun getPrimaryDir() = relativeDir?.split(File.separator)?.firstOrNull()

    fun toMap(): FieldMap {
        return hashMapOf(
            "volumePath" to volumePath,
            "relativeDir" to relativeDir,
        )
    }
}
